pandera/engines/engine.py

```diff
@@ -0,0 +1,157 @@
+import functools
+import inspect
+import warnings
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Type, Union, get_type_hints
+
+import typing_inspect
+
+from pandera.dtypes_ import DataType
+
+
+@dataclass
+class _DtypeRegistry:
+    dispatch: Callable[[Any], DataType]
+    equivalents: Dict[Any, Type[DataType]]
+
+
+class Engine(ABCMeta):
+    """Base Engine.
+
+    Keep a registry of concrete Engines (currently, only pandas).
+    """
+
+    _registry: Dict["Engine", _DtypeRegistry] = {}
+    _base_datatype: Type[DataType]
+
+    def __new__(mcs, name, bases, namespace, **kwargs):
+        base_datatype = kwargs.pop("base_datatype")
+        try:  # allow multiple base datatypes
+            base_datatype = tuple(base_datatype)
+        except TypeError:
+            pass
+        namespace["_base_datatype"] = base_datatype
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+
+        @functools.singledispatch
+        def dtype(obj: Any) -> DataType:
```

Comment:
also consider `native_data_type` here

Reply:
It's not necessarily a native data type. We register "vanilla" DataTypes to lib-specific engines. So I'll rename to data_type :)